components/raftstore-v2/src/operation/command/control.rs

```diff
@@ -0,0 +1,428 @@
+    pub fn record_proposed_admin(&mut self, cmd_type: AdminCmdType, index: u64) {
+        let epoch_state = admin_cmd_epoch_lookup(cmd_type);
+        if !epoch_state.change_conf_ver && !epoch_state.change_ver {
+            return;
+        }
+
+        let conflict_cmd = self.proposed_admin_cmd.iter_mut().rev().find(|cmd| {
+            (epoch_state.check_ver && cmd.epoch_state.change_ver)
+                || (epoch_state.check_conf_ver && cmd.epoch_state.change_conf_ver)
+        });
+        assert!(conflict_cmd.is_none(), "{:?}", conflict_cmd);
+
+        if let Some(cmd) = self.proposed_admin_cmd.back() {
+            assert!(cmd.index < index, "{:?} {}", cmd, index);
+        }
+        self.proposed_admin_cmd
+            .push_back(ProposedAdminCmd::new(cmd_type, epoch_state, index));
+    }
+
+    /// Commit the admin commands.
+    #[inline]
+    pub fn commit_to(&mut self, index: u64, mut on_commit: impl FnMut(&ProposedAdminCmd)) {
+        if self.proposed_admin_cmd.is_empty() {
+            return;
+        }
+
+        for cmd in &mut self.proposed_admin_cmd {
+            if cmd.committed {
+                continue;
+            }
+            if cmd.index <= index {
+                cmd.committed = true;
+                on_commit(cmd);
+                continue;
+            }
+            return;
+        }
+    }
+
+    pub fn advance_apply(&mut self, index: u64, term: u64, region: &metapb::Region) {
+        while !self.proposed_admin_cmd.is_empty() {
+            let cmd = self.proposed_admin_cmd.front_mut().unwrap();
+            if cmd.index <= index {
+                for ch in cmd.delayed_chs.drain(..) {
+                    let mut resp = cmd_resp::new_error(Error::EpochNotMatch(
+                        format!(
+                            "current epoch of region {} is {:?}",
+                            region.get_id(),
+                            region.get_region_epoch(),
+                        ),
+                        vec![region.to_owned()],
+                    ));
+                    cmd_resp::bind_term(&mut resp, term);
+                    ch.report_error(resp);
+                }
+            } else {
+                break;
+            }
+            self.proposed_admin_cmd.pop_front();
+        }
+    }
+
+    #[inline]
+    pub fn enter_prepare_merge(&mut self, prepare_merge_index: u64) {
+        self.pending_merge_index = prepare_merge_index;
+    }
+
+    #[inline]
+    pub fn leave_prepare_merge(&mut self, prepare_merge_index: u64) {
+        if self.pending_merge_index != 0 {
+            assert_eq!(self.pending_merge_index, prepare_merge_index);
+            self.pending_merge_index = 0;
+        }
+    }
+
+    /// Check if there is an on-going split command on current term.
+    ///
+    /// The answer is reliable only when the peer is leader.
+    #[inline]
+    pub fn is_splitting(&self) -> bool {
+        if self.proposed_admin_cmd.is_empty() {
```

Comment:
it's covered below

Reply:
It's just a shortcut for most cases. Iterator will take more instructions to find out it's empty.